### PR TITLE
Add base class for generators

### PR DESCRIPTION
**Pull request by matedabis**
Add base class for generators, with frequently used functions.

### CODE REVIEW

**Comment by matedabis** [comment]
This PR and #2 together are working and up to date to each other.

This way generated files look like this:
```
function validate_numeric(test_value, expected_result, false_results) {
var radix = 2 + Math.abs(expected_result % 35);
assert(test_value === expected_result);
assert(test_value.toString(radix) === expected_result.toString(radix));
for (var i in false_results) {
radix = 2 + Math.abs(false_results[i] % 35);
assert(test_value !== false_results[i]);
assert(test_value.toString(radix) !== false_results[i].toString(radix));
}
}

t1 = 1393823298;
t2 = 1773334627;
t3 = 0;
t4 = -1624120051;
t5 = 1273819165;
t6 = -799240456;
t7 = 84806027;
t8 = 1521124651;
t9 = -1464979674;
t10 = 2124382107;
f1 = 1026581450;
f2 = 2130015057;
f3 = 2134190604;
f4 = 255241881;
f5 = 2133467548;
validate_numeric((((((((((t1 & t2) << t3) & t4) & t5) & t6) >> t7) & t8) & t9) | t10)
, 2124382107, [f1, f2, f3, f4, f5]);

...
```

File: src/scripts/generator_base.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    # Function for printing debug messages if the user wants to (-q switch : quiet)
+    def debug(self, message, options):
+        if not options.q:
+            print(message)
+
+    # Function for generating a random false value integer
+    def random_false_int(self, expected_value, MIN, MAX):
+        # Random number wich must not be equal with the expected value
+        if ((random.randint(0, 1) and expected_value != MAX) or expected_value == MIN):
+            N1 = random.randint(math.ceil(expected_value + 1), MAX)
```
Comment (rerobika) [inline src/scripts/generator_base.py]: Use `int()` instead of `math.ceil()`.

File: src/scripts/generator_base.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    file_output = [] # All the test cases generated (without function definition)
+
+    # Function for printing debug messages if the user wants to (-q switch : quiet)
+    def debug(self, message, options):
+        if not options.q:
+            print(message)
+
+    # Function for generating a random false value integer
+    def random_false_int(self, expected_value, MIN, MAX):
+        # Random number wich must not be equal with the expected value
```
Comment (rerobika) [inline src/scripts/generator_base.py]: typo: `wich`

File: src/scripts/generator_base.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        if ((random.randint(0, 1) and expected_value != MAX) or expected_value == MIN):
+            N1 = random.randint(math.ceil(expected_value + 1), MAX)
+        else:
+            N1 = random.randint(MIN, math.floor(expected_value - 1))
```
Comment (rerobika) [inline src/scripts/generator_base.py]: Same as above.

File: src/scripts/generator_base.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        return N1
+
+    # Generating false numbers and returning with an array
+    def generate_false_numbers(self, expected_value, false_values_count, MIN, MAX):
+        false_values = []
+        for j in range(false_values_count):
+            false_values.append(self.random_false_int(expected_value, MIN, MAX))
+        return false_values
+
+    # Creating a string wich contains an array of false values with variable names, only no. of false values are needed
```
Comment (rerobika) [inline src/scripts/generator_base.py]: Typo again: `wich`

Comment (rerobika) [reply src/scripts/generator_base.py]: Please rephrase this sentence!

File: src/scripts/generator_base.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    def create_false_numbers_array_string(self, false_values_count):
+        false_values = "["
+        for i in range(false_values_count):
+            false_values += "%s" % (_constants.FALSE_RESULT_VAR_NAME.format(NUMBER = i + 1))
+            if (i != false_values_count - 1):
+                false_values += ", "
+        false_values += "]"
+        return false_values
```
Comment (rerobika) [inline src/scripts/generator_base.py]: You could use something like this:

`' '.join ('[%s]'% ','.join([str(x) for x in range(1,10)]))`

This generates: `[ 1 , 2 , 3 , 4 , 5 , 6 , 7 , 8 , 9 ]`.

File: src/scripts/generator_base.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    def create_declarations(self, values, var_name):
+        declaration = ""
+        if isinstance(values, list):
+            for i in range(len(values)): # For eg: a0 = 38473837;
+                declaration += var_name.format(NUMBER = i + 1)
+                declaration += " = %s;\n" % (str(values[i]))
+        else:
+            declaration += var_name.format(NUMBER = "")
+            declaration += " = %s;\n" % (str(values))
+        return declaration
```
Comment (rerobika) [inline src/scripts/generator_base.py]: This kind of similar to the function above. Please try to apply the suggestion to this function as well.

File: src/scripts/generator_base.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+    # Creates the call of the needed function, with as many false values as needed
+    def create_function_call(self, test_value, expected_value, false_values, function):
+        # Calling function and returning with string wich includes function calling
```
Comment (rerobika) [inline src/scripts/generator_base.py]: Ditto.

**Review by rerobika — CHANGES_REQUESTED** [review]

File: src/scripts/generator_base.py (hunk 6 above)
Comment (matedabis) [reply src/scripts/generator_base.py]: Is it okay like this?
```
    def create_declarations(self, values, var_name):
        declaration = ""
        if isinstance(values, list):
            for i in range(1, len(values) + 1): # For eg: a0 = 38473837;
                 declaration += "%s = %s;\n" % (var_name.format(NUMBER = i), str(values[i - 1]))
        else: # If we have only one value no number added to the variable name, for eg: a = 38473837;
            declaration = "%s = %s;\n" % (var_name.format(NUMBER = ""), (str(values)))
        return declaration
```
I could'nt manage to do it that way, and I think in this case it is more readable this way. The code became 2 lines shorter even by this. But in the other case it was useful.

Comment (rerobika) [reply src/scripts/generator_base.py]: Better! But in this function we create a "list" from declarations, and store it in a concatenated string. How about storing the list instead? e.g { "f1" : "5", "f2" : "10, ... }

File: src/scripts/generator_base.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+            print(message)
+
+    # Function for generating a random false value integer
+    def random_false_int(self, expected_value, MIN, MAX):
+        # Random number which must not be equal with the expected value
+        if ((random.randint(0, 1) and expected_value != MAX) or expected_value == MIN):
+            N1 = random.randint(int(expected_value + 1), MAX)
+        else:
+            N1 = random.randint(MIN, int(expected_value - 1))
+        return N1
```
Comment (rerobika) [inline src/scripts/generator_base.py]: Please use the following structure:
```py
if condition:
  return true_result
return false_result
```

**Review by rerobika — CHANGES_REQUESTED** [review]

**Comment by matedabis** [comment]
Corrected according to the suggestions, and synchronised #2 and #11 with this.

File: src/scripts/generator_base.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+sys.path.append(path)
+import validate # Function
+from constants import _constants
+
+# Generator base class, other generator classes will intherit from this class
+class Generator:
+    test_case = "" # Actual test case with validate function call
+    file_output = [] # All the test cases generated (without function definition)
+
+    # Function for printing debug messages if the user wants to (-q switch : quiet)
```
Comment (rerobika) [inline src/scripts/generator_base.py]: `printing debug messages (-q: quiet)` is more than enough

File: src/scripts/generator_base.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    def debug(self, message, options):
+        if not options.q:
+            print(message)
+
+    # Function for generating a random false value integer
```
Comment (rerobika) [inline src/scripts/generator_base.py]: Same as above + `generate random false value`

File: src/scripts/generator_base.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    def random_false_int(self, expected_value, MIN, MAX):
```
Comment (rerobika) [inline src/scripts/generator_base.py]: `random_false_int` -> `generate_false_value`

File: src/scripts/generator_base.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+        # Random number which must not be equal with the expected value
+        if ((random.randint(0, 1) and expected_value != MAX) or expected_value == MIN):
```
Comment (rerobika) [inline src/scripts/generator_base.py]: Add a symbolic name for `random.randint(0, 1)` (what it does exactly)

File: src/scripts/generator_base.py [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+            return random.randint(int(expected_value + 1), MAX)
+        return random.randint(MIN, int(expected_value - 1))
+
+    # Generating false numbers and returning with an array
```
Comment (rerobika) [inline src/scripts/generator_base.py]: `Generating false values into an array`

File: src/scripts/generator_base.py [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+    def generate_false_numbers(self, expected_value, false_values_count, MIN, MAX):
+        false_values = []
+        for j in range(false_values_count):
+            false_values.append(self.random_false_int(expected_value, MIN, MAX))
+        return false_values
+
+    # Creating a string which contains the array of false values
```
Comment (rerobika) [inline src/scripts/generator_base.py]: `Creating a string which contains the array of false values` -> `Stringify an array of false values`

File: src/scripts/generator_base.py [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+    def create_false_numbers_array_string(self, false_values_count):
+        return ('[%s]'% ', '.join([_constants.FALSE_RESULT_VAR_NAME.format(NUMBER = i) for i in range(1,
+        false_values_count + 1)]))
+
+    # Create a string for the declaration(s) of the value(s) in the dictionary (each) as a variable
```
Comment (rerobika) [inline src/scripts/generator_base.py]: In general: `Create a string`-> `Stringify`

Comment (rerobika) [reply src/scripts/generator_base.py]: `Stringify a declaration dictionary`

File: src/scripts/generator_base.py [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+    def create_declarations_string(self, declarations_dictionary):
+        declaration_string = ""
+        for key, value in declarations_dictionary.items():  # For eg: a0 = 38473837;
```
Comment (rerobika) [inline src/scripts/generator_base.py]: `E.g.:`

File: src/scripts/generator_base.py [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+             if isinstance (value, str): # If it is a string, we have to add apostrophes
```
Comment (rerobika) [inline src/scripts/generator_base.py]: `Quote string values`

File: src/scripts/generator_base.py [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+                 declaration_string += "%s = '%s';\n" % (str(key), str(value))
+             else:
+                 declaration_string += "%s = %d;\n" % (str(key), value)
+        return declaration_string
+
+    # Create a dictionary from the value(s) given
```
Comment (rerobika) [inline src/scripts/generator_base.py]: `from the given value(s)`

File: src/scripts/generator_base.py [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+    def create_declarations_dict(self, values, var_name):
+        declarations_dictionary = {}
+        if isinstance(values, list):
+            for i in range(1, len(values) + 1):
+                declarations_dictionary[var_name.format(NUMBER = i)] = values[i - 1]
+        else:
+            declarations_dictionary[var_name.format(NUMBER = "")] = values
+        return declarations_dictionary
+
+    # Creates the call of the needed function, with as many false values as needed
```
Comment (rerobika) [inline src/scripts/generator_base.py]: `Construct a test case.`

File: src/scripts/generator_base.py [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+    def create_function_call(self, test_value, expected_value, false_values, function):
+        # Calling function and returning with string which includes function calling
```
Comment (rerobika) [inline src/scripts/generator_base.py]: remove this line.

File: src/scripts/templates/validate.py [hunk 21]
```diff
@@ -12,3 +12,7 @@
     }
 }
 '''
+
+validate_numeric_header = '''validate_numeric({EXPRESSION}, {EXPECTED_RESULT}, {FALSE_VALUES});
+
```
Comment (rerobika) [inline src/scripts/templates/validate.py]: Remove the new line from the template.

File: src/scripts/constants.py [hunk 22]
```diff
@@ -26,3 +26,6 @@ class _constants:
     bitmin = - bitmax - 1       # The min value of random number for bitwise operations
     bitmax_exposant = 31        # Maximum number for safely shifting an integer in python to
                                 # be precise for JS 32 bit numbers
+
+    TEST_VALUE_VAR_NAME = '''t{NUMBER}'''    # The name of the test variable(s) in the .js files
```
Comment (rerobika) [inline src/scripts/constants.py]: Why these aren't class properties?

**Review by rerobika — CHANGES_REQUESTED** [review]

**Review by rerobika — APPROVED** [review]

**Review by loki04 — APPROVED** [review]
LGTM

**Pull request merged**